opentracing-spring-cloud/src/main/java/io/opentracing/contrib/spring/cloud/TracerUtils.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright 2017-2018 The OpenTracing Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package io.opentracing.contrib.spring.cloud;
+
+import io.opentracing.ScopeManager;
+import io.opentracing.Span;
+import io.opentracing.SpanContext;
+import io.opentracing.Tracer;
+import io.opentracing.propagation.Format;
+
+import org.springframework.beans.factory.BeanFactory;
+
+public class TracerUtils {
+
+  private TracerUtils() {
+  }
+
+  /**
+   * This method returns a proxy that will lazily retrieve the Tracer
+   * bean when first used. This is to avoid premature resolution of
+   * the Tracer bean during evaluation of BeanPostProcessors, which
+   * would subsequently prevent the Tracer bean itself from being
+   * post processed.
+   *
+   * @param beanFactory The bean factory
+   * @return The tracer proxy
+   */
+  public static Tracer getTracer(BeanFactory beanFactory) {
+    return new TracerProxy(beanFactory);
+  }
+
+  public static class TracerProxy implements Tracer {
```

Comment:
I am not sure whether this is the right way how to deal with this. Could we rather autowire BeanFactory and use it to get the tracer?

Reply:
I see you have changed some code to use `BeanFactory`, so is this wrapper still necessary?

Reply:
In those locations direct use of the `BeanFactory` to retrieve the `Tracer` caused the same issue. The only approach that seemed to work was lazily accessing the tracer bean via the proxy.

Reply:
There has to be a better solution to this problem, this looks like a workaround rather than a solution

Reply:
Agree. Will try to spend some more time on it soon, but if anyone has any other ideas feel free to suggest.

Reply:
@alesj @gytis @geoand do you have any thoughts on this?

Reply:
@pavolloffay I will check it out and let you know if have any suggestions

Reply:
@objectiser Have you tried adding Spring's `@Lazy` annotation to the places where the `Tracer` is autowired? It's purpose is to provide the kind of functionality you have manually implemented

Reply:
No hadn't tried that - will give it a go, thanks.

Reply:
I'm curious to hear how it goes :)